javarosa/src/main/java/org/javarosa/xpath/expr/XPathFuncExpr.java

```diff
@@ -720,6 +700,28 @@ public static Double toNumeric(Object o) {
     }
 
     /**
+     * The xpath spec doesn't recognize scientific notation, or +/-Infinity when converting a
+     * string to a number
+     */
+    private static boolean checkForInvalidNumericOrDatestringCharacters(String s) {
+        for (int i = 0; i < s.length(); i++) {
+            char c = s.charAt(i);
+            if (c != '-' && c != '.' && (c < '0' || c > '9')) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Double attemptDateConversion(String s) {
+        if (s.length() == 0) {
+            // This check is needed because toDate() returns a string if given the empty string
+            throw new XPathTypeMismatchException();
+        }
+        return toNumeric(toDate(s));

```

Comment:
Hm, since toDate() here isn't guaranteed to return a date, I think we should probably manually check the return type. You caught one of the special cases above, but any others that got added to the toDate() method may not be mirrored back here, which would be a tough infinite loop to catch.